src/radical/entk/tools/darshan.py

```diff
@@ -0,0 +1,173 @@
+
+__copyright__ = 'Copyright 2024, The RADICAL-Cybertools Team'
+__license__   = 'MIT'
+
+import glob
+
+from typing import Optional, Dict, List, Union
+
+import radical.utils as ru
+
+from .. import Pipeline, Stage, Task
+
+DARSHAN_LOG_DIR = '%(sandbox)s/darshan_logs'
+
+_darshan_activation_cmds = None
+_darshan_env             = None
+_darshan_runtime_root    = None
+
+
+# ------------------------------------------------------------------------------
+#
+def cache_darshan_env(darshan_runtime_root: Optional[str] = None,
+                      modules: Optional[List[str]] = None,
+                      env: Optional[Dict[str, str]] = None) -> None:
+
+    global _darshan_runtime_root
+
+    if _darshan_runtime_root is None:
+        if (darshan_runtime_root and
+                not darshan_runtime_root.startswith('$') and
+                not darshan_runtime_root.startswith('/')):
+            raise RuntimeError('Darshan root directory should be set with '
+                               'either env variable or an absolute path '
+                               f'(provided path: {darshan_runtime_root})')
+        _darshan_runtime_root = darshan_runtime_root or '$DARSHAN_RUNTIME_ROOT'
+
+    global _darshan_activation_cmds
+    global _darshan_env
+
+    if _darshan_activation_cmds is None:
+
+        _darshan_activation_cmds = []
+        for module in modules or []:
+            _darshan_activation_cmds.append(f'module load {module}')
+        for k, v in (env or {}).items():
+            _darshan_activation_cmds.append(f'export {k.upper()}="{v}"')
+
+        _darshan_env = ru.env_prep(pre_exec_cached=_darshan_activation_cmds)
+
+
+# ------------------------------------------------------------------------------
+# decorator to enable darshan for function that generates Pipeline, Stage, Task
+def with_darshan(func,
+                 darshan_runtime_root: Optional[str] = None,
+                 modules: Optional[List[str]] = None,
+                 env: Optional[Dict[str, str]] = None):
+    def wrapper(*args, **kwargs):
+        return enable_darshan(func(*args, **kwargs),
+                              darshan_runtime_root=darshan_runtime_root,
+                              modules=modules,
+                              env=env)
+    return wrapper
+
+
+# ------------------------------------------------------------------------------
+#
+def enable_darshan(pst: Union[Pipeline, Stage, Task, List[Pipeline]],
+                   darshan_runtime_root: Optional[str] = None,
+                   modules: Optional[List[str]] = None,
+                   env: Optional[Dict[str, str]] = None
+                   ) -> Union[Pipeline, Stage, Task]:
+
+    if not pst:
+        raise ValueError('PST object is not provided')
```

Comment:
`pst` is not an optional parameter, so this error should never happen?

Reply:
this is just to catch empty (e.g., `[]`, `{}`, `''`) or `None` values